File: src/brs/web/api/http/handler/GetTransactionBytes.java
package brs.web.api.http.handler;

import brs.Blockchain;
import brs.Transaction;
import brs.TransactionProcessor;
import brs.util.Convert;
import brs.web.api.http.ApiServlet;
import brs.web.api.http.common.LegacyDocTag;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import javax.servlet.http.HttpServletRequest;

import static brs.web.api.http.common.JSONResponses.*;
import static brs.web.api.http.common.Parameters.TRANSACTION_PARAMETER;

public final class GetTransactionBytes extends ApiServlet.JsonRequestHandler {

  private final Blockchain blockchain;
  private final TransactionProcessor transactionProcessor;

  public GetTransactionBytes(Blockchain blockchain, TransactionProcessor transactionProcessor) {
    super(new LegacyDocTag[] {LegacyDocTag.TRANSACTIONS}, TRANSACTION_PARAMETER);
    this.blockchain = blockchain;
    this.transactionProcessor = transactionProcessor;
  }

  @Override
  protected
  JsonElement processRequest(HttpServletRequest req) {

    String transactionValue = req.getParameter(TRANSACTION_PARAMETER);
    if (transactionValue == null) {
      return MISSING_TRANSACTION;
    }

    long transactionId;
    Transaction transaction;
    try {
      transactionId = Convert.parseUnsignedLong(transactionValue);
    } catch (RuntimeException e) {
      return INCORRECT_TRANSACTION;
    }

    transaction = blockchain.getTransaction(transactionId);
    JsonObject response = new JsonObject();
    if (transaction == null) {
      transaction = transactionProcessor.getUnconfirmedTransaction(transactionId);
      if (transaction == null) {
        return UNKNOWN_TRANSACTION;
      }
    } else {
      response.addProperty("confirmations", blockchain.getHeight() - transaction.getHeight());
    }

    response.addProperty("transactionBytes", Convert.toHexString(transaction.getBytes()));
    response.addProperty("unsignedTransactionBytes", Convert.toHexString(transaction.getUnsignedBytes()));

    return response;
  }

}

File: src/main/java/brs/http/GetTransactionBytes.kt
package brs.http


import brs.Blockchain
import brs.Transaction
import brs.TransactionProcessor
import brs.http.JSONResponses.INCORRECT_TRANSACTION
import brs.http.JSONResponses.MISSING_TRANSACTION
import brs.http.JSONResponses.UNKNOWN_TRANSACTION
import brs.http.common.Parameters.TRANSACTION_PARAMETER
import brs.util.parseUnsignedLong
import brs.util.toHexString
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import javax.servlet.http.HttpServletRequest

internal class GetTransactionBytes(private val blockchain: Blockchain, private val transactionProcessor: TransactionProcessor) : APIServlet.JsonRequestHandler(arrayOf(APITag.TRANSACTIONS), TRANSACTION_PARAMETER) {
    internal override fun processRequest(request: HttpServletRequest): JsonElement {
        val transactionValue = request.getParameter(TRANSACTION_PARAMETER) ?: return MISSING_TRANSACTION

        val transactionId: Long
        var transaction: Transaction?
        try {
            transactionId = transactionValue.parseUnsignedLong()
        } catch (e: RuntimeException) {
            return INCORRECT_TRANSACTION
        }

        transaction = blockchain.getTransaction(transactionId)
        val response = JsonObject()
        if (transaction == null) {
            transaction = transactionProcessor.getUnconfirmedTransaction(transactionId)
            if (transaction == null) {
                return UNKNOWN_TRANSACTION
            }
        } else {
            response.addProperty("confirmations", blockchain.height - transaction.height)
        }

        response.addProperty("transactionBytes", transaction.bytes.toHexString())
        response.addProperty("unsignedTransactionBytes", transaction.unsignedBytes.toHexString())

        return response
    }
}
